feat: 151 pemmdb conventionals

daniel-rdt opened the pull request:
Closes #151.

## Changes proposed in this Pull Request
This PR proposes to add TYNDP conventional generation technologies to the model using PEMMDB capacities, must-runs and availabilities. The addition of the conventional generation technologies will build on the previous PEMMDB PRs #128 and #115 and shall implement the following TYNDP generation technologies:

- Nuclear pp
- Gas pp
- Hard coal pp
- Lignite pp
- Hydrogen pp
- Light oil pp
- Heavy oil pp
- Oil shale pp

There is a high level of granularity in the available PEMMDB data with technology carriers (gas, hard coal, lignite, etc) being further separated into multiple different power plant categories. To account for this and keep the implemented technologies easy to grasp, the PR proposes to include a configuration switch that allows to simplify and group together the capacities of different power plant categories according to a pre-determined mapping defined in `data/tyndp_technology_map.csv`.

## Tasks

- [x] Add conventional generation techs to the network with TYNDP technology names and default PyPSA-Eur tech assumptions
- [x] Add grouping option for conventional techs
- [x] Add H2 fuel cell and CCGT with correct name and as non-extendable assets to the network
- [x] Add function `group_tyndp_conventionals` to group together PEMMDB input data as per pre-determined mapping from  `data/tyndp_technology_map.csv`
- [x] Add PEMMDB capacities, must-runs and availabilities to the technologies in the network (for nuclear power plants this includes trajectories)
- [x] Drop the conventional capacities from the previous planning horizon in `add_brownfield`
- [x] Add release note (and other docs)
- [x]  Fix capacities as MWth 

## Workflow
Conventional power plants are per default attached in `prepare_sector_network` in `add_generation`. The implementation will introduce a new section to this function to add the conventional technologies. The following major changes will be made:

- (new) `add_thermal_generation_tyndp`: New function called inside of `add_generation` that adds non-extendable conventional thermal generation technologies from TYNDP. It will iterate over all specified `electricity:tyndp_conventional_carriers` from the config file. The function will only add the required components for the specified technologies but will not add any capacities or must-run conditions yet.
- (new) configuration option `electricity:pemmdb_capacities:group_thermal_techs`: New  configuration switch to group together the capacities and must-runs of different power plant categories of the same carrier as long as the efficiency assumptions do not differ
- (new) `group_tyndp_conventionals`: New rule that groups PEMMDB capacities and profiles of conventional thermal power plants together as per pre-determined mapping from  `data/tyndp_technology_map.csv`. The rule takes in PEMMDB capacities and profiles and outputs a grouped version of the same.
- (new) `update_costs_tyndp`: New intermediate function in `process_cost_data.py` that updates the cost assumptions table by mapping TYNDP carriers to existing default PyPSA-Eur entries and adding the new entries as a copy. Once TYNDP specific technology and cost assumptions are added via [`overwrite_costs.csv`](https://github.com/PyPSA/pypsa-eur/pull/1752#event-20681517447), this function will need to be reverted
- (new) `add_existing_pemmdb_capacities`: New function that will be called on the top level and that will collectively add all of the PEMMDB capacities and must-runs in one place. Inside there will be a differentiation between the addition of the different technology groups, starting in this PR with the conventional thermal power plant capacities. The function takes a pypsa network and adds the PEMMDB capacities in place.
- (new) `add_conventional_thermal_capacities`: New function to add conventional thermal capacities and must-runs to a given PyPSA network. The PEMMDB conventional thermal capacities are given in net generating capacity (MWel) and thus need to be converted to MWth first.
- New section in `add_brownfield` that drops the conventional capacities from the previous planning horizon as PEMMDB capacities are given cummulatively and in PyPSA-Eur they are added as seperate assets for each planning horizon.

<img width="2314" height="1326" alt="image" src="https://github.com/user-attachments/assets/9bbef565-2779-4622-8421-f00bd8ea55b3" />

## Open issues
- What default PyPSA-Eur technology and cost assumptions should we map to the `gas-conv` type? (Note: OCGT and CCGT gas power plants are a separate type)
- Currently missing input data on technology assumptions for PEMMDB technologies (efficiencies, cost)
- For now, it is needed to keep extendable OCGT generators in the model to be feasible as not all TYNDP generation technologies are yet added to the model otherwise.
- There remains a marginal mismatch for three of the included technologies:
  - Oil: Oil shale new for Estonia (EE) is not included in the data of the TYNDP Report and Visualization platform, hence there is less Oil shale in EE in 2030 and no Oil shale in EE in 2040
  - Gas: For Spain (ES) Other Non-RES are counted towards methane power plants in the TYNDP Report and Visualization platform. In this implementation it is not yet included.
  - Coal & other fossils: It is not entirely clear what is included in the `Coal & other fossils` category as the values also differ between the Report and the visualization platform.
 
## Notes
Benchmarking the newly included PEMMDB capacities against reported values yields:
**2030**
<img width="1052" height="729" alt="image" src="https://github.com/user-attachments/assets/fa9b6424-c58f-4995-b44a-9bb93b9789ba" />
**2040**
<img width="1052" height="734" alt="image" src="https://github.com/user-attachments/assets/68de1d08-86b4-4e3b-b216-5f1616a043c6" />

## Checklist

- [x] I tested my contribution locally and it works as intended.
- [x] Code and workflow changes are sufficiently documented.
- [x] Changes in configuration options are added in `config/config.default.yaml`.
- [x] Changes in configuration options are documented in `doc/configtables/*.csv`.
- [x] Changes in configuration options are added in `config/test/*.yaml`.
- [x] Open-TYNDP SPDX license header added to all touched files.
- [x] A release note `doc/release_notes.rst` is added.
- [x] Major features are documented with up-to-date information in `README` and `doc/index.rst`.

lisazeyen commented on scripts/add_brownfield.py:
```diff
@@ -443,6 +447,84 @@ def update_dynamic_ptes_capacity(
+    None
+        This function updates the network in place and does not return a value.
+    """
+    logger.info(
```
it is a bit unclear that we are still tracking cumulative installed capacities also from the previous years, better replace with

"logger.info(
    "Remove cumulative TYNDP conventional capacities from previous planning horizon "
    "and replace with cumulative capacities from new planning horizon."
)"

lisazeyen commented on scripts/prepare_sector_network.py:
```diff
@@ -592,6 +595,34 @@ def update_wind_solar_costs(
+    # Add CCS assumptions as intermediate copy since there is no direct mapping atm
+    costs.loc["gas-ccgt-ccs"] = costs.loc["CCGT"]
+    costs.loc["coal-ccs"] = costs.loc["coal"]
+    costs.loc["lignite-ccs"] = costs.loc["lignite"]
```
costs should be higher for CCS, you can find some assumptions e.g. for ccgt [here](https://github.com/open-energy-transition/open-tyndp/blob/5c290b69e5ed2d35fe0f22f0ce4b53dd20977367/scripts/prepare_sector_network.py#L1179-L1185)

daniel-rdt commented on scripts/prepare_sector_network.py:
```diff
@@ -1380,11 +1411,288 @@ def cycling_shift(df, steps=1):
+                p_nom_extendable=False,
+                capital_cost=costs.at[generator, "efficiency"]
+                * costs.at[generator, "capital_cost"]  # NB: fixed cost is per MWel
+                + costs.at["biomass CHP capture", "capital_cost"]
```
Ah nice, didn't know about this. So, essentially, we can use the assumptions about `cement capture` cost and capture rate for gas-ccgt-cc? But in contrast to [here](https://github.com/open-energy-transition/open-tyndp/blob/5c290b69e5ed2d35fe0f22f0ce4b53dd20977367/scripts/prepare_sector_network.py#L1179-L1185), we probably still want to use gas CO2 intensity instead of methanol co2.

daniel-rdt commented on scripts/prepare_sector_network.py:
```diff
@@ -1503,6 +1813,187 @@ def add_generation(
+
+            # Enable expansion so that pathway supersedes the PEMMDB capacity
+            # TODO: verify how to treat NT scenario
+            n.links.loc[tech_i, "p_nom_extendable"] = True
```
In fact, all of the given trajectories for Nuclear are fixed as `p_nom_min` == `p_nom_max`, i.e. they are giving fixed transition pathways for nuclear via the trajectories. I wanted to keep it flexible though for the case that the inputs change here and `p_nom_min` != `p_nom_max`. What I could do is only setting p_nom_extendable=True for where `p_nom_min` != `p_nom_max` which for our current input data will be none. Otherwise, we manually set `p_nom` = `p_nom_min` = `p_nom_max`.

lisazeyen replied:
Yes, sounds good. Have an extra check and if `p_nom_min==p_nom_max` we can set the `p_nom` and make them non extendable

daniel-rdt commented:
Thanks @lisazeyen for the very helpful review. I've addressed your comments and @tgilon will take over tomorrow for a second round of review before we can hopefully merge the PR before Monday.

I agree with your comment on clearing up some confusion around existing capacities. Tbh, this is sometimes still a bit confusing in the upstream workflow to me (also given some changes that we introduced in our OET fork). Maybe we can brainstorm a good general solution to this at the hackathon. I opened an [issue](https://github.com/open-energy-transition/open-tyndp/issues/210) for this.

tgilon commented on doc/configtables/electricity.csv:
```diff
@@ -29,6 +29,7 @@ everywhere_powerplants,--,"Any subset of {nuclear, oil, OCGT, CCGT, coal, lignit
 ,,, 
 conventional_carriers,--,"Any subset of {nuclear, oil, OCGT, CCGT, coal, lignite, geothermal, biomass}","List of conventional power plants to include in the model from ``resources/powerplants_s_{clusters}.csv``. If an included carrier is also listed in ``extendable_carriers``, the capacity is taken as a lower bound."
 tyndp_conventional_carriers,--,"Any subset of {gas, coal, lignite, nuclear, oil-light, oil-heavy, oil-shale}",List of TYNDP conventional power plants to include in the model.
+group_tyndp_conventionals,bool,"{true, false}",Whether to group together different TYNDP conventional technologies according to a predetermined mapping specified in ``data/tyndp_technology_map.csv``
```
```suggestion
group_tyndp_conventionals,bool,"{true, false}",Whether to group together different TYNDP conventional technologies according to a predetermined mapping specified in ``data/tyndp_technology_map.csv``, as long as the assumptions do not differ.
```

tgilon commented on rules/solve_myopic.smk:
```diff
@@ -76,6 +76,8 @@ def input_profile_tech_brownfield_pecd(w):
 
 rule add_brownfield:
     params:
+        sector=config_provider("sector"),
```
It's an overkill to add `sector` only for two options: `hydrogen_fuel_cell` and `hydrogen_turbine`. I would add these directly as parameters.

daniel-rdt replied:
good point

tgilon commented on scripts/prepare_sector_network.py:
```diff
@@ -7619,11 +8153,71 @@ def add_import_options(
+        tyndp_trajectories = pd.read_csv(snakemake.input.tyndp_trajectories)
+
+    # Extract TYNDP conventional carriers
+    tyndp_conventional_carriers = snakemake.params.conventional_carriers_tyndp
```
```suggestion
    tyndp_conventional_carriers = snakemake.params.tyndp_conventional_carriers
```

tgilon commented on scripts/prepare_sector_network.py:
```diff
@@ -592,6 +615,64 @@ def update_wind_solar_costs(
             )
 
 
+def update_costs_tyndp(
```
I think that this belongs to `prepare_cost_data:prepare_costs`

tgilon commented on scripts/prepare_sector_network.py:
```diff
@@ -592,6 +615,64 @@ def update_wind_solar_costs(
+    """
+    Update technology and cost assumptions for TYNDP technologies based on TYNDP to PyPSA-Eur default carrier mapping.
+
+    costs : pd.DataFrame
```
```suggestion
    Parameters
    ----------
    costs : pd.DataFrame
```

tgilon commented on scripts/prepare_sector_network.py:
```diff
@@ -1380,11 +1461,280 @@ def cycling_shift(df, steps=1):
+    Group conventional carriers based on specified columns.
+    """
+    # Define aggregation strategies for specific columns
+    AGG_STRATEGIES = {
```
If capitalised, it suggests that it's a global variable. You should define it as such.

tgilon commented on scripts/prepare_sector_network.py:
```diff
@@ -1380,11 +1461,280 @@ def cycling_shift(df, steps=1):
+    }
+
+    return (
+        df.groupby(groupby, dropna=False)
```
Grouping the profiles is slow. Have you looked  for any performance improvements yet?

tgilon commented on scripts/prepare_sector_network.py:
```diff
@@ -1380,11 +1461,280 @@ def cycling_shift(df, steps=1):
+            p_max_pu=lambda df: df.p_max / df.p_nom,
+        )
+        .drop(columns=["p_min", "p_max", "p_nom", "p_nom_profiles"], errors="ignore")
+    )
```
```suggestion
        .assign(
            p_max_t=lambda df: df.p_max_t
            + (df.p_nom - df.p_nom_profiles),  # add missing capacities to p_max
            p_min_pu=lambda df: df.p_min_t / df.p_nom,
            p_max_pu=lambda df: df.p_max_t / df.p_nom,
        )
        .drop(columns=["p_min_t", "p_max_t", "p_nom", "p_nom_profiles"], errors="ignore")
    )
```

daniel-rdt replied:
Done

tgilon commented on scripts/prepare_sector_network.py:
```diff
@@ -7619,11 +8153,71 @@ def add_import_options(
+        ).first()
+
+    # Get list of conventional thermal technologies
+    tyndp_conventional_dict, tyndp_conventional_thermals = (
```
Would it be possible to only have one function for preparing mappings?

tgilon reviewed (commented):
@daniel-rdt I haven't finished the review yet. However, here is a first round of comments. I find the changes a bit difficult to navigate, they add even more functions and generation capacities. Unfortunately, I can't suggest  (yet) anything to improve this.

EDIT: This may be because I haven't followed the first part of the discussions.

tgilon commented on scripts/prepare_sector_network.py:
```diff
@@ -7619,11 +8153,71 @@ def add_import_options(
+            snakemake.input.pemmdb_profiles
+        ).to_dataframe()
+    if tyndp_renewable_carriers:
+        tyndp_trajectories = pd.read_csv(snakemake.input.tyndp_trajectories)
```
`snakemake.input.tyndp_trajectories` is only defined on `tyndp_renewables_carriers`:

```python
tyndp_trajectories=branch(
    config_provider("electricity", "tyndp_renewable_carriers"),
    resources("tyndp_trajectories.csv"),
),
```
        
 However, nuclear could be defined without `tyndp_renewables_carriers`.

tgilon commented on scripts/prepare_sector_network.py:
```diff
@@ -1503,6 +1855,212 @@ def add_generation(
+    pemmdb_capacities : pd.DataFrame
+        Dataframe containing all PEMMDB capacities.
+    pemmdb_profiles : pd.DataFrame
+        Dataframe containing all PEMMDB must-run and availability profiles.
```
```suggestion
        DataFrame containing all PEMMDB must-run and availability profiles.
```

tgilon commented on scripts/prepare_sector_network.py:
```diff
@@ -1380,11 +1461,280 @@ def cycling_shift(df, steps=1):
+    tyndp_conventional_mapping : pd.DataFrame
+        Mapping from index_carrier to open_tyndp_type.
+    tyndp_conventional_carriers : list[str]
+        List of TYNDP conventional carriers to group
```
```suggestion
        List of TYNDP conventional carriers to group.
```

tgilon commented on scripts/add_brownfield.py:
```diff
@@ -443,6 +445,84 @@ def update_dynamic_ptes_capacity(
+    Remove TYNDP conventional capacities from previous planning horizon network
+    as existing conventional capacities are given as cumulative input.
+
+    n_p : pypsa.Network
```
```suggestion
    Parameters
    ----------
    n_p : pypsa.Network
```

tgilon commented on scripts/prepare_sector_network.py:
```diff
@@ -1503,6 +1855,212 @@ def add_generation(
+        links_rm = (
+            n.links.loc[tech_i].query("p_nom_extendable == False and p_nom == 0").index
+        )
+        n.remove("Link", links_rm)
```
Like it! Shall we generalise this? This is probably something we want to apply to more than just conventionals.

daniel-rdt replied:
I already opened an issue for the hackathon [here](https://github.com/open-energy-transition/open-tyndp/issues/209).

tgilon replied:
Tks

tgilon commented on config/benchmarking.default.yaml:
```diff
@@ -111,10 +117,16 @@ benchmarking:
         H2 turbine: hydrogen
+        h2-ccgt: hydrogen
+        h2-fuel-cell: hydrogen
         coal: coal + other fossil
```
```suggestion
        coal: coal + other fossil
        coal+ccs: coal + other fossil
```

tgilon commented on config/benchmarking.default.yaml:
```diff
@@ -253,11 +271,17 @@ benchmarking:
+        gas-ccgt: methane&hydrogen
+        gas-ocgt: methane&hydrogen
+        gas-conv: methane&hydrogen
         coal: coal&lignite&oil
```
```suggestion
        coal: coal&lignite&oil
        coal+ccs: coal&lignite&oil
```

tgilon reviewed (commented):
Firstly, thank you @daniel-rdt ! My previous review did not acknowledge all the good and hard work already done to structure and implement this PR. It's great to see that we have almost all of the correct NT numbers. As I was not involved in developing this PR, I am discovering everything at once. Please excuse me if my comments overlap with things that have already been decided.

As always, I've made two types of comments. Some are about styles and structure. The second group is more about the general approach.

My key points are:
- I think we can improve the structure slightly. Now that I know the new code better, however, I can navigate it more easily.
- I'm not yet convinced by the approach chosen for the myopic optimisation. Removing the capacities from the previous planning horizon in `add_brownfield` is not what we are used to do.
- The CI needs to be fixed.
- The function `_group_conv` must be refactored. The current version is not performant on my machine.
- I'm not sure whether a dedicated function to add generations is necessary (i.e.: `add_thermal_generation_tyndp`)
- You will probably need to create another issue to track the remaining work for DE and GA.

daniel-rdt commented on scripts/prepare_sector_network.py:
```diff
@@ -1380,11 +1461,280 @@ def cycling_shift(df, steps=1):
+        how="left",
+    ).assign(
+        p_min=lambda df: df.p_min_pu * df.p_nom,
+        p_max=lambda df: df.p_max_pu * df.p_nom,
```
Done

daniel-rdt commented on scripts/prepare_sector_network.py:
```diff
@@ -1380,11 +1461,280 @@ def cycling_shift(df, steps=1):
+
+    # Group capacities
+    capacities_grouped = _group_capacities(
+        pemmdb_capacities, tyndp_conventional_carriers, tyndp_conventional_mapping
```
Addressed with https://github.com/open-energy-transition/open-tyndp/pull/195/commits/831017f96cba8fd8c5e348327dba151980ae6877

daniel-rdt reviewed (commented):
Thanks @tgilon for all the comments. I have not yet addressed all of the minor comments but I have now implemented a restructuring of the grouping into a separate rule which should definitely make the code in prepare_sector_network leaner and more readable. I also aligned the cost updating to the new structure. I will tag you once I've addressed all of the remaining comments

daniel-rdt commented on scripts/add_brownfield.py:
```diff
@@ -443,6 +445,84 @@ def update_dynamic_ptes_capacity(
     ].values
 
 
+def extract_conventional_thermals_tyndp(
```
Addressed with https://github.com/open-energy-transition/open-tyndp/pull/195/commits/579d30830238ef7653066133f4f01dabf9aa9590

tgilon commented on scripts/prepare_sector_network.py:
```diff
@@ -2056,11 +1815,11 @@ def add_existing_pemmdb_capacities(
-    # Attach existing conventional thermal capacities
+    # Add existing conventional thermal capacities to already attached conventional technologies
     if tyndp_conventional_thermals:
         nuclear_trajectories = trajectories.query(
```
Thanks for the reminder!

lisazeyen reviewed (approved):
Great! Thanks @daniel-rdt ready to be merged from my side as well!
